### ISSUE TITLE
TransformerEncoder is not causal

### ISSUE THREAD

**Issue by bigheary** [pytorch/tutorials#2478]
### Add Link

https://pytorch.org/tutorials/beginner/transformer_tutorial.html

![image](https://github.com/pytorch/tutorials/assets/11831785/285e0fed-1f34-419d-935b-029d43414c37)

for language modeling， src_mask should be mask future words

### Describe the bug

is there anything wrong?

### Describe your environment

 colab

cc @pytorch/team-text-core @Nayef211 @sekyondaMeta @svekars @carljparker @NicolasHug @kit1980 @subramen

**Comment by jorendorff** [comment]
Yes, this has to be a bug. I tried using the resulting model to do some basic predictions and it was never able to give a prediction for the next word that made any sense, despite reaching a perplexity of 2.9 on the validation and test data. Without a mask, the model every word except the last (the one that isn't present in the input) should be easy to predict flawlessly.

**Comment by jorendorff** [comment]
You can [play with it here](https://colab.research.google.com/drive/1ozIEWPtbf3dJakqL6YqjCvg5t6K0GgAX?usp=sharing) (a copy of the notebook with a cell added at the end for making predictions)

**Comment by RustyGrackle** [comment]
@bigheary  Can you tell me more specifically, why you think src_mask should be mask future words (tgt_mask) ??

**Comment by jorendorff** [comment]
The text of the tutorial says:

> Along with the input sequence, a square attention mask is required because the self-attention layers in `nn.TransformerDecoder` are only allowed to attend the earlier positions in the sequence. For the language modeling task, any tokens on the future positions should be masked.

**Comment by jorendorff** [comment]
The purpose of the mask is to prevent the model from seeing the token it's supposed to be predicting. Without a mask, the task is trivial and the model doesn't learn.

Other tutorials also include this mask. Andrej Karpathy's video, for example, lingers for 15 minutes on the subject,[starting here](https://www.youtube.com/watch?v=kCc8FmEb1nY&t=2533s), and then later comes back and [discusses it again](https://www.youtube.com/watch?v=kCc8FmEb1nY&t=4455s), saying "you have to mask ... so that nodes from the future never talk with the past, because they would give away the answer". (He points out that non-predictive applications don't need the mask, but the tutorial at issue here begins "This is a tutorial on training a model to predict the next word[...].")

**Comment by ahoblitz** [comment]
Added a `generate_square_subsequent_mask` with 30 epochs to get the following based on some of the discussion here
<img width="1361" alt="Screenshot 2023-11-07 at 7 40 55 PM" src="https://github.com/pytorch/tutorials/assets/5953164/3f31d016-1a78-40b5-bdeb-97339da56738">

Let me know if this follows the spirit of the conversation or not
